### ISSUE TITLE
Not found certificate matching SNI IP Address

### ISSUE THREAD

**Issue by symflood** [haproxy/haproxy#332]
## Output of `haproxy -vv` and `uname -a`

```
# haproxy -vv

HA-Proxy version 2.0.7 2019/09/27 - https://haproxy.org/
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -fno-strict-aliasing -Wdeclaration-after-statement -fwrapv -Wno-format-truncation -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-old-style-declaration -Wno-ignored-qualifiers -Wno-clobbered -Wno-missing-field-initializers -Wno-implicit-fallthrough -Wno-stringop-overflow -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1

Feature list : +EPOLL -KQUEUE -MY_EPOLL -MY_SPLICE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED -REGPARM -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -VSYSCALL +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -MY_ACCEPT4 +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL -SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=8).
Built with OpenSSL version : OpenSSL 1.1.1d  10 Sep 2019
Running on OpenSSL version : OpenSSL 1.1.1d  10 Sep 2019
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.5
Built with network namespace support.
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with PCRE2 version : 10.33 2019-04-16
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with the Prometheus exporter as a service

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTX        side=FE|BE     mux=H2
              h2 : mode=HTTP       side=FE        mux=H2
       <default> : mode=HTX        side=FE|BE     mux=H1
       <default> : mode=TCP|HTTP   side=FE|BE     mux=PASS

Available services :
	prometheus-exporter

Available filters :
	[SPOE] spoe
	[COMP] compression
	[CACHE] cache
	[TRACE] trace

# uname -a

Linux 91ae86efe6a7 3.13.0-160-generic #210-Ubuntu SMP Mon Sep 24 18:08:15 UTC 2018 x86_64 Linux
```

## What's the configuration?

```
global
    log localhost local0 debug
    log localhost local1 debug
    chroot /var/lib/haproxy
    ssl-default-bind-options ssl-min-ver TLSv1.2 no-tls-tickets
    tune.ssl.default-dh-param 4096
    stats socket /run/haproxy/admin.sock mode 660 level admin expose-fd listeners
    stats timeout 30s
    user haproxy
    group haproxy

defaults
    log global
    option dontlognull
    mode http
    option httplog
    timeout connect 10s
    timeout client 60s
    timeout server 60s
    maxconn 3000

frontend https
    bind *:443 ssl strict-sni crt /etc/ssl/private

    log global
    option httplog

    mode http
    option forwardfor
    redirect scheme https if !{ ssl_fc }

    acl is_agent_github_hook hdr_sub(user-agent) -i GitHub-Hookshot
    acl is_header_nexus hdr(host) -i nexus.local
    acl is_agent_docker hdr_sub(user-agent) -i Docker

    use_backend smee if is_agent_github_hook
    use_backend nexus if is_header_nexus !is_agent_docker
    use_backend nexus-docker-connector if is_header_nexus is_agent_docker

backend smee
    mode http
    http-request set-header X-Forwarded-Port %[dst_port]
    http-request add-header X-Forwarded-Proto https if { ssl_fc }
    server smee-service smee:3000 check

backend nexus
    mode http
    http-request set-header X-Forwarded-Port %[dst_port]
    http-request add-header X-Forwarded-Proto https if { ssl_fc }
    server nexus-service nexus:8081 check

backend nexus-docker-connector
    mode http
    http-request set-header X-Forwarded-Port %[dst_port]
    http-request add-header X-Forwarded-Proto https if { ssl_fc }
    server nexus-service nexus:18081 check

```

## Steps to reproduce the behavior

Don't know if this a haproxy, configuration or unrelated issue. Maybe I've missinterpeted SNI and SANs.
The only thing I've done are two certificates, that were issued by our internal CA and have set `strict-sni`.

- nexus service have CN `nexus.local` and SAN `DNS:nexus.local` 

```
$ cat nexus.local.pem | openssl x509 -noout -text | grep -A1 -P '(X509v3|CN)'

        Issuer: <stripped>
        Validity
--
        Subject: <stripped> CN=nexus.local
        Subject Public Key Info:
--
        X509v3 extensions:
            X509v3 Subject Alternative Name: 
                DNS:nexus.local
            X509v3 Subject Key Identifier: 
                <stripped>
            X509v3 Authority Key Identifier: 
                keyid:<stripped>
--
            X509v3 CRL Distribution Points: 

--
            X509v3 Key Usage: critical
                Digital Signature, Key Encipherment
--
            X509v3 Extended Key Usage: 
                TLS Web Server Authentication
```

- smee service have CN `smee.local` and SAN `DNS:smee.local, IP Address:106.125.46.76` 

```
$ cat smee.local.pem | openssl x509 -noout -text | grep -A1 -P '(X509v3|CN)'
        Issuer: <stripped>
        Validity
--
        Subject: <stripped> CN=smee.local
        Subject Public Key Info:
--
        X509v3 extensions:
            X509v3 Subject Alternative Name: 
                DNS:smee.local, IP Address:106.125.46.76
            X509v3 Subject Key Identifier: 
                <stripped>
            X509v3 Authority Key Identifier: 
                keyid:<stripped>
--
            X509v3 CRL Distribution Points: 

--
            X509v3 Key Usage: critical
                Digital Signature, Key Encipherment
--
            X509v3 Extended Key Usage: 
                TLS Web Server Authentication
```

## Actual behavior

Haproxy doesnt return the certificate if request is done with IP

```
curl -vvv --cacert /etc/ssl/certs/ca-certificates.crt https://106.125.46.76

*   Trying 106.125.46.76...
* TCP_NODELAY set
* Connected to 106.125.46.76 (106.125.46.76) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: /etc/ssl/certs
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS alert, Server hello (2):
* error:14094458:SSL routines:ssl3_read_bytes:tlsv1 unrecognized name
* stopped the pause stream!
* Closing connection 0
curl: (35) error:14094458:SSL routines:ssl3_read_bytes:tlsv1 unrecognized name
```

However everything is fine if I am using the domain name:

```
curl -vvv --cacert /etc/ssl/certs/ca-certificates.crt https://nexus.local

*   Trying 106.125.46.76...
* TCP_NODELAY set
* Connected to nexus.local (106.125.46.76) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: /etc/ssl/certs
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS Unknown, Certificate Status (22):
* TLSv1.3 (IN), TLS handshake, Unknown (8):
* TLSv1.3 (IN), TLS Unknown, Certificate Status (22):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (IN), TLS Unknown, Certificate Status (22):
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
* TLSv1.3 (IN), TLS Unknown, Certificate Status (22):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.3 (OUT), TLS change cipher, Client hello (1):
* TLSv1.3 (OUT), TLS Unknown, Certificate Status (22):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384
* ALPN, server did not agree to a protocol
* Server certificate:
*  subject: <stripped> CN=nexus.local
*  start date: Jun 19 15:21:39 2019 GMT
...


curl -vvv --cacert /etc/ssl/certs/ca-certificates.crt https://smee.local/                                                                                            [057c802]
*   Trying 106.125.46.76...
* TCP_NODELAY set
* Connected to smee.local (106.125.46.76) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: /etc/ssl/certs
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.3 (IN), TLS Unknown, Certificate Status (22):
* TLSv1.3 (IN), TLS handshake, Unknown (8):
* TLSv1.3 (IN), TLS Unknown, Certificate Status (22):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (IN), TLS Unknown, Certificate Status (22):
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
* TLSv1.3 (IN), TLS Unknown, Certificate Status (22):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.3 (OUT), TLS change cipher, Client hello (1):
* TLSv1.3 (OUT), TLS Unknown, Certificate Status (22):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384
* ALPN, server did not agree to a protocol
* Server certificate:
*  subject: <stripped> CN=smee.local
*  start date: Oct 17 13:05:41 2019 GMT
...
```

## Expected behavior

I am expecting that haproxy should return the correct certificate for SNI with IP address.

## Do you have any idea what may have caused this?

No

## Do you have an idea how to solve the issue?

No

**Comment by wlallemand** [comment]
That's the correct behaviour, the strict-sni option will deliver the SNI only if the domain is matched in the client request. It disables the default certificate, which is delivered with any other domain. Just remove the "strict-sni" option if you still want the default certificate. The default certificate is the first crt on a bind line, if you use a directory I suggest you to add a "crt default.pem" before the "crt directory/"

**Comment by symflood** [comment]
@wlallemand I think you've missunderstood me. The behavior I see only correct for SANs DNS. Certificate was also issued for IP address, that's why I've provided SANs `DNS:smee.local, IP Address:106.125.46.76`. Disabled `strict-sni` will results in haproxy returning incorrect certificate, which it should return according to SNI and SANs metadata.

In the example I have provided there are two certificates, one of which is with CN and SAN/DNS and the other is with CN and SAN/DNS and also SAN/IP address. So, I've thought haproxy chooses certificate incorrectly or it doesn't consider the IP address as valid in SNI or in SANs.

---

The real answer I have found is in [RFC4366: 3.1. Server Name Indication](https://tools.ietf.org/html/rfc4366#section-3.1) or [RFC6066: 3. Server Name Indication](https://tools.ietf.org/html/rfc6066#section-3), both of which states

>    Currently, the only server names supported are DNS hostnames;
   however, this does not imply any dependency of TLS on DNS, and other
   name types may be added in the future (by an RFC that updates this
   document).  TLS MAY treat provided server names as opaque data and
   pass the names and types to the application.

>   "HostName" contains the fully qualified DNS hostname of the server,
   as understood by the client.  The hostname is represented as a byte
   string using UTF-8 encoding [UTF8], without a trailing dot.

>   If the hostname labels contain only US-ASCII characters, then the
   client MUST ensure that labels are separated only by the byte 0x2E,
   representing the dot character U+002E (requirement 1 in Section 3.1
   of [IDNA] notwithstanding).  If the server needs to match the
   HostName against names that contain non-US-ASCII characters, it MUST
   perform the conversion operation described in Section 4 of [IDNA],
   treating the HostName as a "query string" (i.e., the AllowUnassigned
   flag MUST be set).  Note that IDNA allows labels to be separated by
   any of the Unicode characters U+002E, U+3002, U+FF0E, and U+FF61;
   therefore, servers MUST accept any of these characters as a label
   separator.  If the server only needs to match the HostName against
   names containing exclusively ASCII characters, it MUST compare ASCII
   names case-insensitively.

 >  Literal IPv4 and IPv6 addresses are not permitted in "HostName".

So it more looks like inconsistency between SNI and SAN extensions standards and the issue not related to the haproxy but to TLS implementation (openssl client doesn't follow the standard and allows IP4/IP6 SNIs to be generated as HostName).
